components/merkledb/benches/schema_patterns.rs

```diff
@@ -0,0 +1,327 @@
+struct Transaction {
+    value: u64,
+    _payload: [u8; 32],
+}
+
+impl BinaryValue for Transaction {
+    fn to_bytes(&self) -> Vec<u8> {
+        bincode::serialize(self).unwrap()
+    }
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Result<Self, failure::Error> {
+        bincode::deserialize(bytes.as_ref()).map_err(From::from)
+    }
+}
+
+impl_object_hash_for_binary_value! { Transaction }
+
+trait ExecuteTransaction {
+    fn execute<T: Access>(fork: T, transaction: &Transaction)
+    where
+        T::Base: RawAccessMut;
+}
+
+struct EagerSchema<T: Access> {
+    // Accessed once per transaction.
+    transactions: MapIndex<T::Base, Hash, Transaction>,
+    // Hot index / group are accessed `DIVISORS.len()` times per transaction.
+    hot_index: ProofMapIndex<T::Base, u64, Hash>,
+    hot_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    // Cold index / group are accessed once per ~10 transactions.
+    cold_index: ProofMapIndex<T::Base, u64, Hash>,
+    cold_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    // Accessed once per ~`COLD_DIVISOR` transactions.
+    other_cold_index: KeySetIndex<T::Base, u64>,
+}
+
+impl<T: Access> EagerSchema<T> {
+    fn new(access: T) -> Self {
+        Self {
+            transactions: Restore::restore(access.clone(), "transactions".into()).unwrap(),
+            hot_index: Restore::restore(access.clone(), "hot_index".into()).unwrap(),
+            hot_group: Restore::restore(access.clone(), "hot_group".into()).unwrap(),
+            cold_index: Restore::restore(access.clone(), "cold_index".into()).unwrap(),
+            cold_group: Restore::restore(access.clone(), "cold_group".into()).unwrap(),
+            other_cold_index: Restore::restore(access.clone(), "other_cold_index".into()).unwrap(),
+        }
+    }
+}
+
+impl<T: Access> EagerSchema<T>
+where
+    T::Base: RawAccessMut,
+{
+    fn execute(&mut self, transaction: &Transaction) {
+        self.transactions
+            .put(&transaction.object_hash(), *transaction);
+
+        // Access hot index and group a few times.
+        for &divisor in DIVISORS {
+            let group_id = transaction.value % divisor;
+            let mut list_in_group = self.hot_group.get(&group_id);
+            list_in_group.push(transaction.value);
+            self.hot_index.put(&group_id, list_in_group.object_hash());
+
+            // Cold index / group are accessed only a fraction of the time.
+            if group_id == 0 {
+                let cold_group_id = transaction.value % COLD_DIVISOR;
+                let mut list_in_group = self.cold_group.get(&cold_group_id);
+                list_in_group.push(transaction.value);
+                self.cold_index
+                    .put(&cold_group_id, list_in_group.object_hash());
+            }
+        }
+
+        if transaction.value % COLD_CHANCE == 0 {
+            self.other_cold_index.insert(transaction.value);
+        }
+    }
+}
+
+enum EagerStyle {}
```

Comment:
Just out of curiosity. Should it be a exactly enum rather than unit-like struct ?

Reply:
Probably should be `struct EagerStyle(!)` as per [clippy suggestion](https://rust-lang.github.io/rust-clippy/master/index.html#empty_enum); don't know why clippy missed this.